modules/objdetect/src/qrcode_encoder.cpp

```diff
@@ -268,53 +269,67 @@ int QRCodeEncoderImpl::findVersionCapacity(const int input_length, const int ecc
     return version_index;
 }
 
-bool QRCodeEncoderImpl::estimateVersion(const int input_length, vector<int>& possible_version)
+bool QRCodeEncoderImpl::estimateVersion(const int input_length, EncodeMode mode, vector<int>& possible_version)
 {
     possible_version.clear();
-    if (input_length > version_capacity_database[40].ec_level[ecc_level].encoding_modes[1])
-        return false;
-    if (input_length <= version_capacity_database[9].ec_level[ecc_level].encoding_modes[3])
-    {
-        possible_version.push_back(1);
-    }
-    else if (input_length <= version_capacity_database[9].ec_level[ecc_level].encoding_modes[1])
+
+    CV_Assert(mode != EncodeMode::MODE_AUTO);
+
+    const int mode_index[] = {
+       -1,
+        0, // cv::QRCodeEncoder::EncodeMode::MODE_NUMERIC
+        1, // cv::QRCodeEncoder::EncodeMode::MODE_ALPHANUMERIC
+       -1,
+        2, // cv::QRCodeEncoder::EncodeMode::MODE_BYTE
+       -1,
+       -1,
+       -1,
+        3  // cv::QRCodeEncoder::EncodeMode::MODE_KANJI
+    };
+
+    if (input_length > version_capacity_database[40].ec_level[ecc_level].encoding_modes[mode_index[mode]])
     {
-        possible_version.push_back(1);
-        possible_version.push_back(2);
+        return false;
     }
-    else if (input_length <= version_capacity_database[26].ec_level[ecc_level].encoding_modes[3])
+
+    int version = 40;
+
+    for (; version > 0; --version)
     {
-        possible_version.push_back(2);
+        if (input_length > version_capacity_database[version].ec_level[ecc_level].encoding_modes[mode_index[mode]]) {
+            break;
+        }
     }
-    else if (input_length <= version_capacity_database[26].ec_level[ecc_level].encoding_modes[1])
+
+    if (version < MAX_VERSION)
     {
-        possible_version.push_back(2);
-        possible_version.push_back(3);
+        version += 1;
     }
-    else
+
+    possible_version.push_back(version);
+
+    if (version < MAX_VERSION)
     {
-        possible_version.push_back(3);
+        possible_version.push_back(version + 1);
     }
+
```

Comment:
Two similar if statements `if (version < MAX_VERSION)`

Reply:
Why do not push all versions between `version` and `MAX_VERSION`?